src/pages/library/Showcase.js

```diff
@@ -16,104 +18,337 @@ import * as Style from 'src/libs/style'
+            onClick: () => setOpenState(_.update(section.name, open => !open, openState)),
+            isOpened: openState[section.name]
+          },
+          [
+            ..._.map(label => {
+              const tag = _.toLower(label)
+              return h(Clickable, {
+                style: { display: 'flex', alignItems: 'baseline', margin: '0.5rem 0' },
+                onClick: () => onTagFilter(tag)
+              }, [
+                div({ style: { flex: 1 } }, [label]),
+                div({
+                  style: _.includes(tag, selectedTags) ? styles.hilightedPill : styles.pill
+                }, [_.size(workspacesByTag[tag])])
+              ])
+            }, section.labels)
+          ]
+        )
+    }, sections)
+  ])
+}
 
 const Showcase = () => {
-  const { workspaces } = useWorkspaces()
   const stateHistory = StateHistory.get()
   const [featuredList, setFeaturedList] = useState(stateHistory.featuredList)
+  const [workspacesByTag, setWorkspacesByTag] = useState({})
+  const [sections, setSections] = useState([])
+
+  const [selectedSections, setSelectedSections] = useState([])
+  const [selectedTags, setSelectedTags] = useState([])
+  const [searchFilter, setSearchFilter] = useState()
+  const [sort, setSort] = useState('most recent')
+
+  const sortWorkspaces = Utils.cond(
+    [sort === 'most recent', () => _.orderBy(['created'], ['desc'])],
+    [sort === 'alphabetical', () => _.orderBy(w => _.toLower(_.trim(w.name)), ['asc'])],
+    () => _.identity
+  )
 
   Utils.useOnMount(() => {
     const loadData = async () => {
-      const featuredList = await Ajax().Buckets.getFeaturedWorkspaces()
+      const showcase = await Ajax().Buckets.getShowcaseWorkspaces()
 
-      setFeaturedList(featuredList)
-      StateHistory.update({ featuredList })
+      // Immediately lowercase the workspace tags so we don't have to think about it again.
+      // Also pre-compute lower name and description.
+      const featuredWorkspaces = _.flow([
+        _.map(_.update(['tags', 'items'], _.map(_.toLower))),
+        _.map(workspace => ({ ...workspace, lowerName: _.toLower(workspace.name), lowerDescription: _.toLower(workspace.description) }))
+      ])(showcase)
+
+      const workspacesByTag = _.omitBy(_.isEmpty, groupByFeaturedTags(featuredWorkspaces))
+
+      // Trim items from the sidebar for which there aren't any featured workspaces.
+      const activeTags = _.keys(workspacesByTag)
+      const activeSections = _.flow([
+        _.map(_.update(['labels'], labels => _.intersectionBy(_.toLower, labels, activeTags))),
+        _.map(section => ({ ...section, tags: _.map(_.toLower, section.labels) })),
+        _.remove(section => _.isEmpty(section.labels))
+      ])(sidebarSections)
+
+      setFeaturedList(featuredWorkspaces)
+      setWorkspacesByTag(workspacesByTag)
+      setSections(activeSections)
+
+      // TODO: is this actually doing anything for us?
+      StateHistory.update({ featuredWorkspaces })
```

Comment:
are we leaving this TODO here for now?

Reply:
Hmm... yeah, I'd like to understand this better than I do. I believe the intent of this is a performance optimization, but I don't think it's entirely working. In addition, since there are a few calculations that are happening from this data, we'd either need to include those in state history, change the `useOnMount` logic to continue doing the calculations based on the cached data from state history, or move the calculations into `_.once` or `_.memoize` functions... or some other solution I'm not thinking of.

For now, I'm out of time for working on this project, so unfortunately I'm going to have to leave this TODO alone.

Reply:
FWIW, the intention of these state history saves isn't to prevent a fetch, it's to show old data while the new data is being fetched, so users can start to navigate without waiting.

Reply:
https://user-images.githubusercontent.com/1545444/129941034-ae6e6fbc-55b3-4a5c-aa10-9b1ad280a0f8.mov

Ignore the error notifications and focus on the interaction. When I load the page, you see a spinner. When I navigate back to it, the page loads instantaneously.  This is the type of attention to detail and user empathy that puts Terra UI in the top tier.